src/components/inputs/Combobox/Combobox.tsx

```diff
@@ -0,0 +1,564 @@
+import { useEffect, useRef, useState } from 'react';
+import Fuse from 'fuse.js';
+import CheckSVG from '../../../svg/checkmark--big.svg';
+import DownloadSmallSVG from '../../../svg/download--small.svg';
+import ArrowRightSVG from '../../../svg/arrow--right.svg';
+import styles from './Combobox.scss';
+import { FunctionGeneric } from '../../../common/structures/Generics';
+import { CaretIcon, CloseSmallIcon, SearchIcon } from '../../icons/Icons';
+import BasicInput from '../BasicInput/BasicInput';
+import { IconButton } from '../../buttons/IconButton/IconButton';
+import { Container } from '../../modules/Container/Container';
+import ILocalContainerProps from '../../../common/structures/ILocalContainerProps';
+
+export interface ComboboxOption {
+	disabled?: boolean;
+	download?: boolean;
+	icon?: React.ReactNode;
+	label: string;
+	metadata?: any;
+	optionGroup?: string;
+	secondaryText?: React.ReactNode;
+}
+
+// Guarantees the key of each state.options option will equal option.value
+export interface ComboboxOptionFormatted extends ComboboxOption {
+	value: string;
+}
+
+// Guarantees structure of state.options - we should refactor this to an array of objects instead of an object of objects maybe
+export type ComboboxOptionsFormatted = ComboboxOptionFormatted[];
+export type ComboboxOptions =
+	| ComboboxOptionFormatted[]
+	| string[]
+	| { [value: string]: ComboboxOption }
+	| { [value: string]: string };
+export type ComboboxOptionsLoader = () => Promise<ComboboxOptions>;
+
+export interface ComboboxOptionGroup {
+	label: string;
+	linkText?: string;
+	linkOnClick?: FunctionGeneric;
+}
+export type ComboboxOptionGroups = { [key: string]: ComboboxOptionGroup };
+
+export interface IComboboxProps extends ILocalContainerProps {
+	disabled?: boolean;
+	emptyPlaceholder?: string;
+	loadingOptionsPlaceholder?: string;
+	onChange: FunctionGeneric;
+	options?: ComboboxOptions;
+	optionHeight?: 's' | 'l';
+	inputHeight?: 's' | 'l';
+	optionsLoader?: ComboboxOptionsLoader;
+	optionGroups?: ComboboxOptionGroups;
+	placeholder?: string;
+	striped?: boolean;
+	value?: string;
+	id: string;
+	noResultsMessage?: string;
+	invalid?: boolean;
+	invalidMessage?: string;
+}
+
+interface IState {
+	focusedIndex: number;
+	options: ComboboxOptionsFormatted;
+	open: boolean;
+	optionsLoaded: boolean;
+	value: IComboboxProps['value'];
+	filter: string;
+	shouldFilter: boolean;
+}
+
+const Combobox = (props: IComboboxProps) => {
+	const {
+		container,
+		className,
+		style,
+		disabled,
+		emptyPlaceholder,
+		loadingOptionsPlaceholder,
+		onChange,
+		options = {},
+		optionHeight = 's',
+		inputHeight = 'l',
+		optionsLoader,
+		optionGroups,
+		placeholder,
+		striped,
+		value,
+		id,
+		invalid,
+		invalidMessage,
+		noResultsMessage = 'None found. Try a new search term.',
+	} = props;
+
+	const formatOption = (
+		option: ComboboxOption,
+		optionValue: ComboboxOptionFormatted['value']
+	): ComboboxOptionFormatted => {
+		return {
+			value: optionValue,
```

Comment:
If you don't format this to add an explicit optionValue, where would the key value (that you are trying to guarantee) of state.options come from? I was having a little trouble following.

Reply:
Not sure I understand your question

Reply:
I'm basically wondering why it's necessary to have a formatted and unformatted version of the component. Are they mutually exclusive in their usage?

Reply:
Ah, it's to allow a flexible interface for the options prop. Like you can pass an array, and object, or an already formatted object. Internally, the component logic only ever uses formatted options so that we can rely on the data structure.